### ISSUE TITLE
The new nightly version seems to have some bugs

### ISSUE THREAD

**Issue by xzores** [odin-lang/Odin#2679]
## Context

I upgraded to the newest version of Odin (nightly) and my old code would not compile.

Odin: dev-2023-07-nightly:5ac7fe45
OS:   Windows 11 Home Basic (version: 22H2), build 22621.1992
CPU:  13th Gen Intel(R) Core(TM) i9-13900HX
RAM:  32507 MiB

## Expected Behavior

I would expect my code to compile, the compiler gives me errors that make no sense.

## Current Behavior

The compiler gives me errors like:
```
Cannot assign value 'entries[index]' of type '
Octant($Holding=Empty, $Index_type=u8) (package octree)
' to '
Octant($Holding=Empty, $Index_type=u8) (package octree)
' in assignment
```

```
Cannot assign value '_get_sort_encoded(Holding, Index_type, "prune")' of type '
proc(Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8)) -> bool
' to '
proc(Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8)) -> bool
' in procedure argument
```

```
No procedures or ambiguous call for procedure group 'append' that match with the given arguments
        append(&entries, e);
        ^~~~~^
        Given argument types: (^[dynamic]Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8))
```

This used to work, and they types match as you can see, but it says they don't

## Failure Information (for bugs)

I currently have my directory setup like this:
```
my_game
	-> octree package (includes 2 files a test file and the octree data structure)
	-> network package (has 5 files, 1 is a test file)
	-> ...
	-> ...
```
They bugs only occure in my octree package, and they only occure when I and compiling from my game package (which include the octree package). If I "cd octree" and then compile my tests in there, they is no compliants.

### Steps to Reproduce

I could not tell you as a this only seems to be a bug in some packages and not others, and it only seems to be a bug with `odin run .` or `odin build .` , but not `odin test .`.

EDIT:

As you can see from this failure log, it can acctually compile, sometimes but it just failes like 80% and the time.
I havec experienced this before, but never as freqently as in this new release.
### Failure Logs

```

PS D:\github\Voxel_game> odin run .
D:/github/Voxel_game/Octree/Octree.odin(46:5) Error: No procedures or ambiguous call for procedure group 'append' that match with the given arguments
        append(&entries, Octant(Holding, Index_type){encoding = enc, dat ...
        ^~~~~^
        Given argument types: (^[dynamic]Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8))
Did you mean to use one of the following:
        runtime.append_elem        :: proc(array: ^$T/[dynamic]$E, arg: E, loc := #caller_location) -> (n: int, err: Allocator_Error)            at C:/Program Files/Odin/core/runtime/core_builtin.odin(410:1)
        runtime.append_elems       :: proc(array: ^$T/[dynamic]$E, args: ..E, loc := #caller_location) -> (n: int, err: Allocator_Error)         at C:/Program Files/Odin/core/runtime/core_builtin.odin(438:1)
        runtime.append_elem_string :: proc(array: ^$T/[dynamic]$E/u8, arg: $A/string, loc := #caller_location) -> (n: int, err: Allocator_Error) at C:/Program Files/Odin/core/runtime/core_builtin.odin(473:1)

D:/github/Voxel_game/Octree/Octree.odin(130:13) Error: No procedures or ambiguous call for procedure group 'append' that match with the given arguments
        append(&entries, e);
        ^~~~~^
        Given argument types: (^[dynamic]Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8))
Did you mean to use one of the following:
        runtime.append_elem        :: proc(array: ^$T/[dynamic]$E, arg: E, loc := #caller_location) -> (n: int, err: Allocator_Error)            at C:/Program Files/Odin/core/runtime/core_builtin.odin(410:1)
        runtime.append_elems       :: proc(array: ^$T/[dynamic]$E, args: ..E, loc := #caller_location) -> (n: int, err: Allocator_Error)         at C:/Program Files/Odin/core/runtime/core_builtin.odin(438:1)
        runtime.append_elem_string :: proc(array: ^$T/[dynamic]$E/u8, arg: $A/string, loc := #caller_location) -> (n: int, err: Allocator_Error) at C:/Program Files/Odin/core/runtime/core_builtin.odin(473:1)

D:/github/Voxel_game/Octree/Octree.odin(139:35) Error: Cannot assign value '_get_sort_encoded(Holding, Index_type, "prune")' of type 'proc(Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8)) -> bool' to 'proc(Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8)) -> bool' in procedure argument
        ... ies[:], _get_sort_encoded(Holding, Index_type, "prune")); ...
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^

PS D:\github\Voxel_game> cd .\Octree\
PS D:\github\Voxel_game\Octree> odin run .  
D:/github/Voxel_game/Octree/Test.odin(65:27) 
Error: 'o' of type 'Octree(Empty, u8)' has no field 'entries'
        testing.expect(t, len(o.entries) == 0);
                              ^
        Suggestion: Did you mean?
                entries

D:/github/Voxel_game/Octree/Octree.odin(1:2) Error: Undefined entry point procedure 'main'

         ^
PS D:\github\Voxel_game\Octree> odin run .  
D:/github/Voxel_game/Octree/Test.odin(65:27) Error: 'o' of type 'Octree(Empty, u8)' has no field 'entries'
        testing.expect(t, len(o.entries) == 0);
                              ^
        Suggestion: Did you mean?
                entries

D:/github/Voxel_game/Octree/Octree.odin(1:2) Error: Undefined entry point procedure 'main'

         ^
PS D:\github\Voxel_game\Octree> odin run .
D:/github/Voxel_game/Octree/Test.odin(65:27) Error: 'o' of type 'Octree(Empty, u8)' has no field 'entries'
        testing.expect(t, len(o.entries) == 0);
                              ^
        Suggestion: Did you mean?
                entries

D:/github/Voxel_game/Octree/Octree.odin(1:2) Error: Undefined entry point procedure 'main'

         ^
PS D:\github\Voxel_game\Octree> odin run .
D:/github/Voxel_game/Octree/Octree.odin(1:2) Error: Undefined entry point procedure 'main'

         ^
PS D:\github\Voxel_game\Octree> odin test .        
[Package: octree]
[Test: test_1]
[test_1 : SUCCESS]
[Test: test_2]
_bin_search
[test_2 : SUCCESS]
----------------------------------------
2/2 SUCCESSFUL
PS D:\github\Voxel_game\Octree> 
```

**Comment by xzores** [comment]
This is most likely a race condition causing this

**Comment by JesseRMeyer** [comment]
Are you willing to share your Octree package source?  We can't diagnose nor fix this from the symptoms alone.

**Comment by Kelimion** [comment]
It might also be helpful to run `git bisect` and see if you can narrow it down to a commit before which the project compiled without errors. The 5ac7fe45 commit doesn't seem to read on the dynamic array errors you say you're getting.

**Comment by xzores** [comment]
I did try to go a few weeks back, they where all buggy, and this bug has been in there for ever since it got multi threaded. It just more pronounced lately.

Also it is really bad when first installing odin, and then it gets better (mostly), sometimes it gets worse

**Comment by xzores** [comment]
"Are you willing to share your Octree package source? We can't diagnose nor fix this from the symptoms alone."
Sure, how would you like it shared?

**Comment by JesseRMeyer** [comment]
If the package is a single source file, then a gist would work, otherwise, uploading to a public github repo would be the next easiest.

**Comment by colrdavidson** [comment]
Curious to know, if you run Odin with `-thread-count:1`, do you still see the issue?

**Comment by xzores** [comment]
Yes, even when the thread-count is one, the issue is still there.

**Comment by xzores** [comment]
usually gives this error:

```
testing.expect(t, len(o.entries) == 0);
                              ^
        Suggestion: Did you mean?
                entries
 ```

**Comment by xzores** [comment]
Also this has been tired on 3 diffrent PC's they all have same behavior.

**Comment by xzores** [comment]
> If the package is a single source file, then a gist would work, otherwise, uploading to a public github repo would be the next easiest.

Github:
https://github.com/xzores/Linear_octree

**Comment by xzores** [comment]
Great! are you developing the compiler? or do we need someone else to fix it?

**Comment by JesseRMeyer** [comment]
Unfortunately I am the least familiar with the internals of the parapoly subsystem, so I think this will come down to a gB fix in the end.

**Comment by xzores** [comment]
It differs a lot from PC to PC, my PC is it only like 30-50% of the time, but of my friends PC he can hit up to 20 times before it compiles correctly. 
I have intel 32 threads, he has AMD ryzen 8 threads

**Comment by rick-masters** [comment]
The first type of problem is that there is a concurrency problem with checking structure fields:

```
Error: 'o' of type 'Octree(Empty, u8)' has no field 'entries'
        testing.expect(t, len(o.entries) == 0);
```

This problem is a duplicate of Issue #2463 and is fixed by PR #3326, although there is a complication as described in Issue #3327.

The second type of problem is that there are some assignment statements or calls where types are checked for
compatibility but this fails even though they should be the same types.

```
Cannot assign value '_get_sort_encoded(Holding, Index_type, "prune")' of type '
proc(Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8)) -> bool
' to '
proc(Octant($Holding=Empty, $Index_type=u8), Octant($Holding=Empty, $Index_type=u8)) -> bool
' in procedure argument
```

Similar errors have been reported in Issues #2814, #2949, and #3291.

The root cause is that different threads can create and add the same type twice because they may both call `find_polymorphic_record_entity` around the same time, don't find the type, and then proceed to create and then add the type with `add_polymorphic_record_entity`.

A forthcoming PR resolves that problem.
